src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
-     * would drain the queues, similar to observeOn. It killed the Netflix API when I canaried it. There were two problems:
-     * (1) performance and (2) object allocation overhead causing massive GC pressure. Remember that merge is one of the most
-     * used operators (mostly due to flatmap) and is therefore critical to and a limiter of performance in any application.
-     * 
-     * All subsequent work on this class and the various fast-paths and branches within it have been to achieve the needed functionality
-     * while reducing or eliminating object allocation and keeping performance acceptable.
-     * 
-     * This has meant adopting strategies such as:
-     * 
-     * - ring buffers instead of growable queues
-     * - object pooling
-     * - skipping request logic when downstream does not need backpressure
-     * - ScalarValueQueue for optimizing synchronous single-value Observables
-     * - adopting data structures that use Unsafe (and gating them based on environment so non-Oracle JVMs still work)
-     * 
-     * It has definitely increased the complexity and maintenance cost of this class, but the performance gains have been significant.
-     * 
-     * The biggest cost of the increased complexity is concurrency bugs and reasoning through what's going on.
-     * 
-     * I'd love to have contributions that improve this class, but keep in mind the performance and GC pressure.
-     * The benchmarks I use are in the JMH OperatorMergePerf class. GC memory pressure is tested using Java Flight Recorder
-     * to track object allocation.
+    /**
+     * Creates a new instance of the operator with the given delayError and maxConcurrency settings.
+     * @param delayErrors
+     * @param maxConcurrent the maximum number of concurrent subscriptions or Integer.MAX_VALUE for unlimited
+     * @return
      */
-
-    private OperatorMerge() {
-        this.delayErrors = false;
+    public static <T> OperatorMerge<T> instance(boolean delayErrors, int maxConcurrent) {
+        if (maxConcurrent == Integer.MAX_VALUE) {
+            return instance(delayErrors);
+        }
+        return new OperatorMerge<T>(delayErrors, maxConcurrent);
     }
 
-    private OperatorMerge(boolean delayErrors) {
+    final boolean delayErrors;
+    final int maxConcurrent;
+
+    private OperatorMerge(boolean delayErrors, int maxConcurrent) {
         this.delayErrors = delayErrors;
+        this.maxConcurrent = maxConcurrent;
     }
 
-    private final boolean delayErrors;
-
     @Override
     public Subscriber<Observable<? extends T>> call(final Subscriber<? super T> child) {
-        return new MergeSubscriber<T>(child, delayErrors);
-
+        MergeSubscriber<T> subscriber = new MergeSubscriber<T>(child, delayErrors, maxConcurrent);
+        MergeProducer<T> producer = new MergeProducer<T>(subscriber);
+        subscriber.producer = producer;
+        producer.lazySet(0); // make sure subscriber.producer is visible
```

Comment:
What is this lazySet about? What about visibility would not work if this wasn't done?